gno.land/pkg/integration/testing_integration.go

```diff
@@ -0,0 +1,257 @@
+)
+
+type testNode struct {
+	*node.Node
+	logger      log.Logger
+	nGnoKeyExec uint // Counter for execution of gnokey.
+}
+
+func SetupGnolandTestScript(t *testing.T, txtarDir string) testscript.Params {
+	t.Helper()
+
+	cmd := exec.Command("go", "list", "-m", "-mod=mod", "-f", "{{.Dir}}", "github.com/gnolang/gno")
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+
+	// `gnoRootDir` should point to the local location of the gno repository.
+	// It serves as the gno equivalent of GOROOT.
+	gnoRootDir := strings.TrimSpace(string(out))
+
+	// `gnoHomeDir` should be the local directory where gnokey stores keys.
+	gnoHomeDir := filepath.Join(t.TempDir(), "gno")
+
+	// `gnoDataDir` should refer to the local location where the gnoland node
+	// stores its configuration and data.
+	gnoDataDir := filepath.Join(t.TempDir(), "data")
+
+	var muNodes sync.Mutex
+	nodes := map[string]*testNode{}
+
+	persistWorkDir, _ := strconv.ParseBool(os.Getenv("TESTWORK"))
+	return testscript.Params{
+		TestWork: persistWorkDir,
+		Dir:      txtarDir,
+		Setup: func(env *testscript.Env) error {
+			kb, err := keys.NewKeyBaseFromDir(gnoHomeDir)
+			if err != nil {
+				return err
+			}
+
+			// XXX: Add a command to add custom account.
+			kb.CreateAccount("test1", test1Seed, "", "", 0, 0)
+			env.Setenv("USER_SEED_test1", test1Seed)
+			env.Setenv("USER_ADDR_test1", test1Addr)
+
+			env.Setenv("GNOROOT", gnoRootDir)
+			env.Setenv("GNOHOME", gnoHomeDir)
+			env.Setenv("GNODATA", gnoDataDir)
+
+			return nil
+		},
+		Cmds: map[string]func(ts *testscript.TestScript, neg bool, args []string){
+			"sleep": func(ts *testscript.TestScript, neg bool, args []string) {
+				d := time.Second
+				if len(args) > 0 {
+					var err error
+					if d, err = time.ParseDuration(args[0]); err != nil {
+						ts.Fatalf("unable to parse duration %q: %s", args[1], err)
+					}
+				}
+
+				time.Sleep(d)
+			},
+			"gnoland": func(ts *testscript.TestScript, neg bool, args []string) {
+				muNodes.Lock()
+				defer muNodes.Unlock()
+
+				if len(args) == 0 {
+					tsValidateError(ts, "gnoland", neg, fmt.Errorf("use gnoland [start|stop] command"))
+					return
+				}
+
+				sid := getSessionID(ts)
+
+				var cmd string
+				cmd, args = args[0], args[1:]
+
+				var err error
+				switch cmd {
+				case "start":
+					if _, ok := nodes[sid]; ok {
+						err = fmt.Errorf("node %q already started", sid)
+						break
+					}
+
+					logger := log.NewNopLogger()
+					if persistWorkDir || os.Getenv("LOG_DIR") != "" {
+						logname := fmt.Sprintf("gnoland-%s.log", sid)
+						logger = getTestingLogger(ts, logname)
+					}
+
+					dataDir := filepath.Join(gnoDataDir, sid)
+					var node *node.Node
+					if node, err = execTestingGnoland(t, logger, dataDir, gnoRootDir, args); err == nil {
+						nodes[sid] = &testNode{
+							Node:   node,
+							logger: logger,
+						}
+						ts.Defer(func() {
+							if n := nodes[sid]; n != nil {
```

Comment:
Does this not happen outside of the "mutex guard"?

Reply:
Absolutely, it slipped my mind